### PR TITLE
API-level DSSE signing support

### PR DESCRIPTION
**Pull request by woodruffw**
Adds support for DSSE signature generation (over an in-toto statement) at the API level. Not exposed via the `sigstore` CLI yet.

See #628.

### CODE REVIEW

**Comment by woodruffw** [comment]
CCing @laurentsimon, just so you're aware: this PR will change the `sign(...)` API a bit, so #666 may be affected by it 🙂

File: sigstore/sign.py [hunk 1]
```diff
@@ -174,10 +177,9 @@ def _signing_cert(
 
     def sign(
         self,
-        input_: IO[bytes],
+        input_: IO[bytes] | Statement,
```
Comment (woodruffw) [inline sigstore/sign.py:180]: Flagging for review: this puts a third-part type (in-toto-attestation's `Statement`) in our public API.

Normally I'd be neutral on this, but it has a few other "meh" factors:

1. `in-toto-attestation` is still pre-1.0, meaning that we probably shouldn't lash our stable semver'd public API to it;
2. `Statement` is a thin wrapper for a `protobuf` generated `protobuf.Message` class, which both leaks `protobuf` implementation details into the public API *and* may change soon-ish (see https://github.com/in-toto/attestation/issues/291).

Another option here would be to create our own `sigstore.intoto_or_whatever.Statement` type, which would be a thin constructable shim over `in_toto_attestation.Statement`.

File: sigstore/sign.py [hunk 2]
```diff
@@ -308,31 +331,46 @@ def signer(
         yield Signer(identity_token, self, cache)
 
 
-class SigningResult(BaseModel):
+@dataclass
+class SigningResult:
```
Comment (woodruffw) [inline sigstore/sign.py:335]: Flagging for review: the changes here indicate that we should probably just drop this type entirely and switch to `Bundle` as our "signing result" type. It'd be strictly equivalent and would save us a layer of types.

**Comment by woodruffw** [comment]
For posterity: this is a small demo script that you can use to exercise the new DSSE functionality:

```python
#!/usr/bin/env python

import logging

from in_toto_attestation.v1.resource_descriptor import ResourceDescriptor
from in_toto_attestation.v1.statement import Statement

from sigstore.oidc import Issuer
from sigstore.sign import SigningContext

logging.getLogger().setLevel(logging.DEBUG)

ctx = SigningContext.staging()
with ctx.signer(identity_token=Issuer.staging().identity_token()) as signer:
    stmt = Statement(
        subjects=[
            ResourceDescriptor(
                name="null",
                digest={
                    "sha256": "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855"
                },
            ).pb,
        ],
        predicate={"Data": "", "Timestamp": "2023-12-07T00:37:58Z"},
        predicate_type="https://cosign.sigstore.dev/attestation/v1",
    )
    res = signer.sign(stmt)
    print(res)
```

**Comment by woodruffw** [comment]
R4R but marking as blocked, pending resolution of https://github.com/in-toto/attestation/pull/315

**Comment by tetsuo-cpp** [comment]
Will get onto this at the beginning of next week.

File: sigstore/sign.py (hunk 1 above)
Comment (tetsuo-cpp) [reply sigstore/sign.py:180]: I'm not against a shim, but I don't see this causing a problem in practice. Even if/when they move to betterproto, the way users will construct a `Statement` will be pretty much identical. If the `in-toto-attestation` API completely changes before 1.0 to the point where it somehow becomes painful for our users (maybe their API takes different types?), we'll probably need to make breaking changes to the shim anyway.

File: sigstore/sign.py (hunk 2 above)
Comment (tetsuo-cpp) [reply sigstore/sign.py:335]: I think consolidating to using `Bundle` is a good idea. I'm happy to follow up with a change to do that to keep things simple in this PR.

**Review by tetsuo-cpp — APPROVED** [review]
LGTM!

File: sigstore/sign.py (hunk 2 above)
Comment (jku) [reply sigstore/sign.py:335]: This sounds right

**Review by jku — COMMENTED** [review]
I have nothing to add here but left some high level questions in the issue

File: sigstore/sign.py (hunk 2 above)
Comment (woodruffw) [reply sigstore/sign.py:335]: Thanks both! @tetsuo-cpp, I'll assign you a follow-up for removing `SigningResult`.

**Comment by woodruffw** [comment]
For tracking purposes: our upstream blocker is itself blocked on https://github.com/danielgtaylor/python-betterproto/pull/551

File: sigstore/sign.py (hunk 1 above)
Comment (woodruffw) [reply sigstore/sign.py:180]: That's fair. I'll consider this a non-blocker for now, which means this PR is good to go.

**Comment by woodruffw** [comment]
This should be good to go -- I'm setting to auto-merge, pending review 🙂

**Comment by woodruffw** [comment]
> * adding this signing capability without defining the verification side is a bit sketchy (even if I understand that it might not make sense to implement inside sigstore-python): I expect after this PR even the sigstore-python CLI might happily "verify" a bundle with an intoto statement?  This is not entirely wrong but could maybe give the wrong impression to the user (implying that something in the statement has been verified, when it has not).

Yeah, I'm not sure what to do about this -- IMO sigstore-python itself should never attempt to verify these statements (that belongs at a higher-up level), so maybe this is an API documentation problem? Something like a big warning on `verify(...)` saying that the API only verifies the signature itself modulo `VerificationPolicy`, which only concerns the Sigstore bits and not any arbitrary metadata in the payload itself?

> I think merging as is (after conflict fixes) should be ok though if the above items are in plans.

Yeah, I'd like to do an E2E test here once we have verification in place. I could probably hack it together as an integration-style test with another client for verification, but I figure it's not worth the effort versus merging now + working on a verification test in the follow-up.

**Comment by jku** [comment]
> Yeah, I'd like to do an E2E test here once we have verification in place. I could probably hack it together as an integration-style test with another client for verification

Maybe I'm missing something but couldn't you do a test that does signing like your demo and then verifies the bundle (leaving a TODO for verifying the intoto statement) -- just to make sure the round trip works fine?

>> I expect after this PR even the sigstore-python CLI might happily "verify" a bundle with an intoto statement?

I wonder if I'm correct here and if we want to do something about this?
* should the CLI **not** succeed when the payload is a intoto statement (until we know how to expose this)
* should there should be some output on CLI when the payload is a intoto statement

I'm really not sure, this is just hand waving

**Comment by woodruffw** [comment]
> Maybe I'm missing something but couldn't you do a test that does signing like your demo and then verifies the bundle (leaving a note that the intoto statement is not verified) -- just to make sure the round trip works fine?

Not with the current APIs -- I've only implemented DSSE signing, not verification. The verification APIs still expect only a "bare" signature, so attempting to verify a bundle containing a DSSE message will fail (probably with an abstruse Pydantic error).

> I wonder if I'm correct here and if we want to do something about this?

Yes, this is why I don't want to expose these APIs via the CLI in the immediate term 🙂 -- my plan was to have the APIs themselves present in the 3.0 series, after which we can iterate on an appropriate CLI design.

**Review by jku — APPROVED** [review]
Okay, sounds good then :+1:

**Pull request merged**